### ISSUE TITLE
Support the W3C webdriver spec

### ISSUE THREAD

**Issue by stof** [minkphp/MinkSelenium2Driver#293]
Support for the JsonWire selenium protocol is fading out in favor of the W3C spec:

- Firefox's geckodriver implements only the W3C protocol
- Safari's driver uses the W3C driver by default (see https://github.com/minkphp/Mink/issues/764)
- Chrome's driver is working on supporting the W3C protocol (they still implement JsonWire for now)
- Edge's driver uses the W3C driver by default in its next release (so same situation than Safari will happen soon: https://docs.microsoft.com/en-us/microsoft-edge/webdriver#webdriver-server-command-line-flags)

We will have some work to do to make this happen.

The instaclick/php-webdriver library that we are using has a beta version of 2.0 which adds support for interacting with a W3C server. But they won't provide an abstraction layer to make the same code work for both variants (see https://github.com/instaclick/php-webdriver/issues/84). It will be up to our code to choose the right way (and we will need https://github.com/instaclick/php-webdriver/issues/86 to be able to know which one we need).
Due to the fact that we expose the webdriver session to userland, we would have to consider this upgrade as a BC break.

Another option could be to migrate to facebook/webdriver, which seems a bit more popular this days (https://packagist.org/?query=webdriver). It is currently adding support for the W3C spec (WIP is at https://github.com/facebook/php-webdriver/pull/560), and does so by abstracting the differences internally).
Doing such migration would also require making it a major release, as we would change the API we expose when exposing the session.

**Comment by stof** [comment]
https://github.com/minkphp/MinkSelenium2Driver/issues/288 was created to refer to the first solution for this topic (but does not describe the implication, and does not seem to care about the W3C support in the PR implementing it)

**Comment by stof** [comment]
#287 should be done first before this work (we should switch to Chrome on Travis, as there is no way to run recent Firefox versions with the JsonWire protocol).

**Comment by stof** [comment]
The more I look at instaclick/php-webdriver 2 "supporting" the  WEC protocol (currently in RC), the more I think we should rather migrate to the facebook package (for which the PR supporting the W3C protocol inside their abstraction is almost ready AFAICT).

**Comment by stof** [comment]
Btw, if we look at https://packagist.org/packages/instaclick/php-webdriver/dependents and the download counts on https://packagist.org/?query=webdriver, I think most of the downloads of `instaclick/php-webdriver` are coming from Mink using it (our driver has 11 176 049 downloads, the library has 11 517 849, so this only means 340k downloads not coming from us, which is 3%).
Among the 12 packages depending on `instaclick/php-webdriver` on Packagist, here is my analysis:
- 1 is us
- 2 are (avandoned) forks of us, (one direct fork, and a ghost-driver which duplicated our codebase instead of depending on us)
- 3 are about locking exact versions of deps in the Drupal ecosystem and so are related to us (as Drupal is using Mink)
- 1 is an abandoned experiment from the Laravel ecosystem, but `laravel/dusk` has been based on the facebook library instead.
- 1 is a fork of the previous experiment
- the remaining 4 look abandoned (last commits either in 2016 or 2014)

So we are the only ones keeping `instaclick/php-webdriver` alive in the open.

**Comment by aik099** [comment]
> https://github.com/minkphp/MinkSelenium2Driver/issues/293#issuecomment-426670405

Unfortunately Facebook Web Driver is now requiring PHP 7 and that was main reason why PR still isn't merged.

**Comment by stof** [comment]
@aik099 as we need to make a new major release of the driver anyway, I'm fine with requiring PHP 7.1+ for it. All older PHP versions will be EOLed by the end of December.

People really wanting to use PHP 5.x could still use Selenium2Driver 1.x and use Chrome (or another browser with a driver supporting the legacy protocol)

**Comment by stof** [comment]
Well, Mink itself will probably bump its own requirement in the future too, but I don't have a roadmap for that (as long as we don't make changes benefiting from it, there is no strong need)

**Comment by nirbhaypatil** [comment]
just wanted to know how to use instaclick/php-webdriver:2.0.0-RC2 with mink and mink extension ?
dependencies mismatch error comes up.

**Comment by stof** [comment]
currently, you cannot do it directly.

**Comment by nirbhaypatil** [comment]
can you suggest other ways to do .

**Comment by nirbhaypatil** [comment]
Sorry about confusion ,my question is 
Can we use
1) mink + minkextension + instaclick/php-webdriver:2.0.0-RC2
OR
2) mink + minkextension + facebook/webdriver 
to execute tests on firefox > 48

**Comment by stof** [comment]
none of that is possible today, which is precisely what this issue is about:

- using facebook/webdriver would involve rewriting the driver entirely, as it is a different library. You cannot just install it and expect it to be used
- installing instaclick/php-webdriver:2.0.0-RC2 will not make it work with recent Firefox, even if you trick composer into accepting it. instaclick/php-webdriver does not abstract the differences between the W3C and JsonWire protocol, but exposes APIs supporting both and expects the calling code to use the right one, which MinkExtension does not do it yet. And we cannot do it based on their RC2 releases:
    - they don't expose whether the session is a W3C or JsonWire one, so we would have to guess it currently (without having the info to detect it reliably): https://github.com/instaclick/php-webdriver/issues/86
    - their timeout handling (one of the few APIs they actually abstract) is broken with W3C: https://github.com/instaclick/php-webdriver/issues/88
    - the error handling is broken with W3C: https://github.com/instaclick/php-webdriver/issues/87

So we are still discussing in which direction we want to rewrite the driver (supporting the W3C spec on top of instaclick/php-webdriver:2 also involves lots of rewriting in the driver).

**Comment by oleg-andreyev** [comment]
@stof recently we've been trying to migrate this library to `facebook/php-webdriver`, mostly because of `syn.js` which bundled into this library and because `instaclick/php-webdriver` is not supported that actively.

So far my results:
Diff: https://github.com/minkphp/MinkSelenium2Driver/compare/master...oleg-andreyev:master
Travis Build (with Chrome): https://travis-ci.org/oleg-andreyev/MinkSelenium2Driver/jobs/462514905
Travis Build (with FF): https://travis-ci.org/oleg-andreyev/MinkSelenium2Driver/jobs/462514904
Travis: https://travis-ci.org/oleg-andreyev/MinkSelenium2Driver/builds/462514900

There are few issues with Chrome eg: Keyboard event, but it it's related to assert of driver-testsuite.
Also there are issues with facebook/webdriver and with FF

**Comment by phil-davis** [comment]
Note: chrome+chromedriver 75 has recently come out and now defaults to `w3c`
So now I think `MinkSelenium2Driver` users need to set `chromeOption` `w3c` to `false`

Also see https://github.com/SeleniumHQ/docker-selenium/issues/922 where I am having this trouble with the latest https://github.com/SeleniumHQ/docker-selenium `3.141.59-palladium` which has chrome 75.

**Comment by alexdesignworks** [comment]
@phil-davis Looks like we were debugging this at the same time :))

This is the full fix in `behat.yml`:
```
  extensions:
    Behat\MinkExtension:
      selenium2:
         capabilities: { "browser": "chrome", "version": "*", "marionette": true, "extra_capabilities": { "chromeOptions": { "w3c": false } } }
```

**Comment by phil-davis** [comment]
@alexdesignworks I have been trying that in https://github.com/owncloud/core/pull/35445
In `BEHAT_PARAMS` I end up with
```
"capabilities": {"marionette":null, "browser": "chrome", "platform": "Linux", "name": "owncloud/core - ", "extra_capabilities": {"chromeOptions":{"w3c":false},"maxDuration":"3600"}}
```
https://drone.owncloud.com/owncloud/core/18090/681

But I get a WebDriver Exception:
```
    When user "user1" logs in using the webUI                                                           # WebUILoginContext::theUserLogsInUsingTheWebUI()
      WebDriver\Exception\CurlExec: Webdriver http error: 400, payload :{"value":{"error":"invalid argument","message":"invalid argument: missing command parameters","stacktrace":"#0 0x55c24baff299 \u003Cunknown>\n"}} in /drone/src/vendor-bin/behat/vendor/instaclick/php-webdriver/lib/WebDriver/Exception.php:155
      Stack trace:
      #0 /drone/src/vendor-bin/behat/vendor/instaclick/php-webdriver/lib/WebDriver/AbstractWebDriver.php(132): WebDriver\Exception::factory(-1, 'Webdriver http ...')
      #1 /drone/src/vendor-bin/behat/vendor/instaclick/php-webdriver/lib/WebDriver/AbstractWebDriver.php(218): WebDriver\AbstractWebDriver->curl('POST', '/click', NULL)
      #2 /drone/src/vendor-bin/behat/vendor/instaclick/php-webdriver/lib/WebDriver/Container.php(224): WebDriver\AbstractWebDriver->__call('click', Array)
      #3 /drone/src/vendor-bin/behat/vendor/behat/mink-selenium2-driver/src/Selenium2Driver.php(781): WebDriver\Container->__call('click', Array)
      #4 /drone/src/vendor-bin/behat/vendor/behat/mink-selenium2-driver/src/Selenium2Driver.php(769): Behat\Mink\Driver\Selenium2Driver->clickOnElement(Object(WebDriver\Element))
      #5 /drone/src/vendor-bin/behat/vendor/behat/mink/src/Element/NodeElement.php(153): Behat\Mink\Driver\Selenium2Driver->click('(//html/.//*[./...')
      #6 /drone/src/tests/acceptance/features/lib/LoginPage.php(71): Behat\Mink\Element\NodeElement->click()
      #7 /tmp/ProxyManagerGeneratedProxy__PM__PageLoginPageGenerated6312256563dff9947346b9792a817253.php(34): Page\LoginPage->loginAs('user1', '1234', 'FilesPage')
```

Any clues are welcome!

**Comment by mnocon** [comment]
@phil-davis We're having the same error with our test suite (after setting w3c:false).
I'll keep investigating, but for now we will stay with 3.141.59-oxygen to unblock tests.

**Comment by alexdesignworks** [comment]
@phil-davis I'm getting the same. That is actually comes from `Selenium2Driver.php` 
```
    private function clickOnElement(Element $element)
    {
        try {
            // Move the mouse to the element as Selenium does not allow clicking on an element which is outside the viewport
            $this->wdSession->moveto(array('element' => $element->getID()));
        } catch (UnknownCommand $e) {
            // If the Webdriver implementation does not support moveto (which is not part of the W3C WebDriver spec), proceed to the click
        }

        $element->click();
    }
```

Click does not have any arguments.

I've tried to explicitly provide the id of the element as an argument, but that triggers a different error

```
      Webdriver http error: 404, payload :{"value":{"error":"unknown command","message":"unknown command: unknown command: session/7fb3f6976b62fa8e2b39886c4775d31e/element/0.4630396792632474-3/click/0.4630396792632474-3","stacktrace":"#0 0x55af4081e299 \u003Cunknown>\n"}} (WebDriver\Exception\CurlExec)
```

**Comment by alexsegura** [comment]
Just faced the same error, like @mnocon I fixed it by rollbacking to Chrome 74. 

I'm thinking about switching to Chrome Headless, but I'm not sure it would fix the problem.

**Comment by individual-it** [comment]
There is a bug in the chrome driver https://bugs.chromium.org/p/chromedriver/issues/detail?id=2943
workarounds in `facebook/php-webdriver` https://github.com/facebook/php-webdriver/pull/640 and `nightwatch` https://github.com/nightwatchjs/nightwatch/pull/2122

**Comment by phil-davis** [comment]
Where should we do a "fix" like https://github.com/facebook/php-webdriver/pull/640
In this `MinkSelenium2Driver` repo, or in https://github.com/instaclick/php-webdriver or?

**Comment by leymannx** [comment]
Temporarily pinning to Chromedriver 74 seems to be the best workaround for now. See https://github.com/acquia/blt/pull/3708/files/9aa7d8cf28025b957f1e628d9c6de6e63a0a5aac#diff-586532737a140c4f8e2079e2c33fc217

**Comment by jimarasim** [comment]
When setting w3c to false, use "goog:chromeOptions" instead of "chromeOptions"
          extra_capabilities:
            "goog:chromeOptions":
              w3c:
                false

Hoping this issue is resolved before Chrome decides to no longer support jsonwire at all.

**Comment by emodric** [comment]
Is it only me, or this started to work on its own in latest versions of Chrome Driver 75 and even in the new v76?

**Comment by phil-davis** [comment]
I am using `selenium/standalone-chrome-debug` in drone-docker. `latest`  of that still fails, and I have to specify `3.141.59-oxygen`

**Comment by emodric** [comment]
Here is a build that works with latest chrome and chrome driver that previously failed: https://travis-ci.org/netgen-layouts/layouts-core/jobs/566143335

Previous failing build: https://travis-ci.org/netgen-layouts/layouts-core/jobs/545643705#L420

**Comment by shwetaneelsharma** [comment]
@alexdesignworks Thanks a ton for this solution https://github.com/minkphp/MinkSelenium2Driver/issues/293#issuecomment-499458269
However, does it work for headless execution too? I am only able to get it working for browser-based execution

**Comment by shwetaneelsharma** [comment]
I have got this working on Chrome 75 in headless mode by using the following configuration:
`extensions:
    Behat\MinkExtension:
        base_url: "your_base_url"
        browser_name: 'chrome'
        goutte: ~
        javascript_session: selenium2
        selenium2:
          wd_host: http://127.0.0.1:4444/wd/hub
          capabilities:
            browser: chrome
            extra_capabilities:
              chromeOptions:
                args: ['--headless', '--disable-gpu']
                w3c: false`

**Comment by jimarasim** [comment]
With Chrome 76.0.3809.100 / Chromedriver Version = 76.0.3809.68 '--headless' only works for me if "chromeOptions:" is "goog:chromeOptions"; otherwise, it still runs headed.

**Comment by marcusanderss** [comment]
The problem is possible 'fix' it by inserting the following 3 lines at the start of the webDriverElement function in Behat, vendor/instaclick/php-webdriver/lib/WebDriver/Container.php:
    protected function webDriverElement($value)
    {
        if (!array_key_exists('ELEMENT', $value)) {
            $value['ELEMENT'] = $value['element-6066-11e4-a52e-4f735466cecf'];
        }

**Comment by robbieaverill** [comment]
Hi everyone, facebook/php-webdriver support for the W3C protocol is nearly ready: https://github.com/facebook/php-webdriver/pull/560#issuecomment-552471707. Would be worth testing an update to use it now? I'm happy to assemble a PR from [@tractorcow's fork (to use facebook/php-webdriver)](https://github.com/silverstripe/MinkFacebookWebDriver) if it would help as well.

**Comment by AntoscencoVladimir** [comment]
> When setting w3c to false, use "goog:chromeOptions" instead of "chromeOptions"
> extra_capabilities:
> "goog:chromeOptions":
> w3c:
> false
> 
> Hoping this issue is resolved before Chrome decides to no longer support jsonwire at all.

Thank you!

**Comment by stof** [comment]
Closing this as we have created the `mink/webdriver-classic-driver` package.
